File: solutions/fortran/collatz-conjecture/2/collatz_conjecture.f90

module collatz_conjecture
  implicit none
contains

  integer function collatz_map(n)
    integer :: n
    if (mod(n,2) == 0) then
      n = n/2
    else 
      n = 3*n + 1
   end if
    collatz_map = n
  end function collatz_map

  integer function steps(i) result(count)
    integer, intent(in) :: i
    integer ::  j
    count = 0
    if (i>0) then
        j = i
       do while (j /= 1) 
          j = collatz_map(j)
          count = count + 1
       end do
    else
      count = -1
    end if
  end function steps

end module collatz_conjecture

File: solutions/free-pascal/collatz-conjecture/1/CollatzConjecture.pas
unit CollatzConjecture;

{$mode ObjFPC}{$H+}

interface

function steps(const number : integer) : integer;

implementation

uses SysUtils;

function steps(const number : integer) : integer;
var count : integer;
var n : integer;
begin
  
  if number<=0 then
      raise Exception.Create('Only positive integers are allowed');
  n := number;
  count := 0;
  while n<>1 do
  begin
    if n mod 2 = 0 then
      n := n div 2
    else
      n := 3*n + 1;
    inc(count);
  end;
  result:=count;
end;

end.
